### PR TITLE
[dynamo] Add kwarg constructor for `call_dict`

### LINKED ISSUE #96459: [torch.compile] Warning using BuiltinVariable.call_dict but not for {}
### 🐛 Describe the bug

I would expect torch.compile to work equally well for dictionaries created using `dict` as for dictionaries created with `{}`, but using `dict` leads to the warning `incorrect arg count <bound method BuiltinVariable.call_dict of BuiltinVariable(dict)> got an unexpected keyword argument 'loss' and no constant handler`.

Is this expected behavior?

Here's a small reproduction:

```
class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.embedding = torch.nn.Embedding(10, 8)

    def forward(self, input_ids):
        outputs = self.embedding(input_ids)
        loss = outputs.mean()
        return dict(loss=loss, outputs=outputs)  # warning
        # return {"loss": loss, "outputs": outputs} ok
 ```
test with
```
model = torch.compile(Model())
model(torch.randint(10, (4, 16)))
```

### Error logs

2.1.0.dev20230307
[2023-03-09 17:17:55,890] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo start tracing forward
[2023-03-09 17:17:55,895] torch._dynamo.variables.builtin: [WARNING] incorrect arg count <bound method BuiltinVariable.call_dict of BuiltinVariable(dict)> got an unexpected keyword argument 'loss' and no constant handler
[2023-03-09 17:17:55,895] torch._dynamo.output_graph: [INFO] Step 2: calling compiler function debug_wrapper
[2023-03-09 17:17:56,774] torch._inductor.compile_fx: [INFO] Step 3: torchinductor compiling FORWARDS graph 0
[2023-03-09 17:17:58,713] torch._inductor.compile_fx: [INFO] Step 3: torchinductor done compiling FORWARDS graph 0
[2023-03-09 17:17:58,713] torch._dynamo.output_graph: [INFO] Step 2: done compiler function debug_wrapper

### Minified repro

_No response_

### Versions

PyTorch version: 2.1.0.dev20230307
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @davidberard98

### PR DESCRIPTION
**Pull request by udbhav1**
Fixes #96459

Tested on the simple repro in the issue but not sure if there's some nuance around `ConstDictVariable` that I have to consider for more complex cases.

Added a test in `test/dynamo/test_repros.py` that checks to ensure no warnings are emitted for the simple case. Ideally this would only check for the specific `incorrect arg count` warning that prompted the original issue, but the code for that is ugly since it probably requires using `assertNoLogs` in a try/catch to check if we have logs at all and then `assertLogs` to check the logs themselves.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -805,8 +805,10 @@ def call_dict_helper(tx, user_cls, arg, **options):
         else:
             raise AssertionError("call_dict_helper with illegal arg")
 
-    def call_dict(self, tx, obj=None):
+    def call_dict(self, tx, obj=None, **kwargs):
         if self.is_supported_call_dict_arg(tx, obj):
+            if kwargs:
+                return self.call_dict_helper(tx, dict, variables.ConstDictVariable(kwargs, dict))
```
Comment (yanboliang) [inline torch/_dynamo/variables/builtin.py:811]: * We should check if ```kwargs``` is supported, you can extend the function ```is_supported_call_dict_arg``` to ```is_supported_call_dict_arg_kwargs```. For unsupported kwargs, we just fall back. In your current implementation, it throws exception for unsupported cases.
* When you building ```variables.ConstDictVariable(kwargs, dict)```, you have to fill the ```source``` and ```mutable_local```, which is used when reconstruction after graph break.

File: test/dynamo/test_repros.py [hunk 2]
```diff
@@ -2625,6 +2626,20 @@ def fn(x):
...
+                super().__init__()
+                self.a = torch.randn(4)
+
+            def forward(self, x):
+                outputs = torch.sin(x) + torch.cos(self.a)
+                return dict(x=x, outputs=outputs)
+
+        with self.assertNoLogs(logger="torch._dynamo", level=logging.WARNING):
+            foo = torch.compile(Foo())
+            foo(torch.randn(4))
```
Comment (yanboliang) [inline test/dynamo/test_repros.py:2641]: Testing if output specific log is not a good idea, let's use if graph break to test this. You can add a test follow & after this:
https://github.com/pytorch/pytorch/blob/b04f86363fce9828b066ee516e78912691129b3f/test/dynamo/test_functions.py#L611-L617

All test in ```dynamo/test_functions.py``` is no graph break.

**Review by yanboliang — CHANGES_REQUESTED** [review]

File: test/dynamo/test_repros.py (hunk 2 above)
Comment (udbhav1) [reply test/dynamo/test_repros.py]: Added `test_call_dict6`, you're right that it's much simpler

File: torch/_dynamo/variables/builtin.py (hunk 1 above)
Comment (udbhav1) [reply torch/_dynamo/variables/builtin.py:811]: 1. What is an example of unsupported kwargs we should fall back on?

2. `call_dict_helper` when called with a `ConstDictVariable` as the `arg` will set `mutable_local`. However, `source` is not set for any `ConstDictVariable` in this file, so I'm not sure what to use for it.

**Comment by yanboliang** [comment]
@udbhav1 I think this is duplicated with #98660.

**Comment by udbhav1** [comment]
all good, ill close this pr then
